CDC panic on deregister downstream

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
<!-- You can run `tikv-server --version` -->

A custom build TiKV: overvenus/tikv@78621ac . Not sure whether the release branch has the same or not.

<details> <summary> TiKV panic log </summary>

```
[2021/04/29 22:02:51.611 +08:00] [INFO] [endpoint.rs:351] ["cdc deregister"] [deregister="Deregister { deregister: \"downstream\", region_id: 614, downstream_id: DownstreamID(1538), conn_id: ConnID(17), err: Some(Sink(Congest)) }"]
[2021/04/29 22:02:51.618 +08:00] [ERROR] [endpoint.rs:1256] ["cdc send scan event failed"] [req_id=6202]
[2021/04/29 22:02:51.618 +08:00] [INFO] [endpoint.rs:351] ["cdc deregister"] [deregister="Deregister { deregister: \"downstream\", region_id: 642, downstream_id: DownstreamID(1552), conn_id: ConnID(17), err: Some(Sink(Congest)) }"]
[2021/04/29 22:02:51.626 +08:00] [ERROR] [endpoint.rs:1256] ["cdc send scan event failed"] [req_id=6191]
[2021/04/29 22:02:51.626 +08:00] [INFO] [endpoint.rs:351] ["cdc deregister"] [deregister="Deregister { deregister: \"downstream\", region_id: 652, downstream_id: DownstreamID(1541), conn_id: ConnID(17), err: Some(Sink(Congest)) }"]
[2021/04/29 22:02:52.208 +08:00] [INFO] [endpoint.rs:682] ["cdc send event failed, full"] [downstream="\"ipv4:172.16.5.33:58526\""] [conn_id=ConnID(11)]
[2021/04/29 22:02:52.208 +08:00] [INFO] [endpoint.rs:682] ["cdc send event failed, full"] [downstream="\"ipv4:172.16.5.33:58128\""] [conn_id=ConnID(9)]
[2021/04/29 22:02:53.208 +08:00] [INFO] [endpoint.rs:682] ["cdc send event failed, full"] [downstream="\"ipv4:172.16.5.33:58018\""] [conn_id=ConnID(1)]
[2021/04/29 22:02:53.208 +08:00] [INFO] [endpoint.rs:682] ["cdc send event failed, full"] [downstream="\"ipv4:172.16.5.33:58582\""] [conn_id=ConnID(15)]
[2021/04/29 22:02:53.208 +08:00] [INFO] [endpoint.rs:682] ["cdc send event failed, full"] [downstream="\"ipv4:172.16.5.33:58528\""] [conn_id=ConnID(12)]
[2021/04/29 22:02:53.208 +08:00] [INFO] [endpoint.rs:682] ["cdc send event failed, full"] [downstream="\"ipv4:172.16.5.33:58128\""] [conn_id=ConnID(9)]
[2021/04/29 22:02:53.805 +08:00] [INFO] [delegate.rs:128] ["cdc send event failed, full"] [req_id=4874] [downstream_id=DownstreamID(231)] [conn_id=ConnID(9)]
[2021/04/29 22:02:53.805 +08:00] [INFO] [endpoint.rs:351] ["cdc deregister"] [deregister="Deregister { deregister: \"region\", region_id: 4, observe_id: ObserveID(327), err: Sink(Congest) }"]
[2021/04/29 22:02:53.805 +08:00] [INFO] [delegate.rs:341] ["cdc met region error"] [error=Sink(Congest)] [region_id=4]
[2021/04/29 22:02:53.813 +08:00] [INFO] [endpoint.rs:490] ["cdc register region"] [downstream_id=DownstreamID(1563)] [req_id=6214] [conn_id=ConnID(15)] [region_id=4]
[2021/04/29 22:02:53.813 +08:00] [INFO] [endpoint.rs:1300] ["cdc downstream is initialized"] [downstream_id=DownstreamID(1563)]
[2021/04/29 22:02:53.814 +08:00] [ERROR] [endpoint.rs:1256] ["cdc send scan event failed"] [req_id=6214]
[2021/04/29 22:02:53.814 +08:00] [INFO] [endpoint.rs:351] ["cdc deregister"] [deregister="Deregister { deregister: \"downstream\", region_id: 4, downstream_id: DownstreamID(1563), conn_id: ConnID(15), err: Some(Sink(Congest)) }"]
[2021/04/29 22:02:54.195 +08:00] [FATAL] [lib.rs:452] ["called `Result::unwrap()` on an `Err` value: Noop"] [backtrace="stack backtrace:\n   0: tikv_util::set_panic_hook::{{closure}}\n             at components/tikv_util/src/lib.rs:451\n   1: std::panicking::rust_panic_wi
th_hook\n             at library/std/src/panicking.rs:595\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at library/std/src/panicking.rs:497\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at library/std/src/sys_common/ba
cktrace.rs:141\n   4: rust_begin_unwind\n             at library/std/src/panicking.rs:493\n   5: core::panicking::panic_fmt\n             at library/core/src/panicking.rs:92\n   6: core::result::unwrap_failed\n             at library/core/src/result.rs:1355\n   7: core::r
esult::Result<T,E>::unwrap\n             at /home/stn/.rustup/toolchains/nightly-2021-04-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1037\n   8: cdc::endpoint::Endpoint<T>::on_deregister\n             at /home/stn/tikv/components/cdc/src/en
dpoint.rs:375\n   9: <cdc::endpoint::Endpoint<T> as tikv_util::worker::pool::Runnable>::run\n             at /home/stn/tikv/components/cdc/src/endpoint.rs:1288\n  10: tikv_util::worker::pool::Worker::start_with_timer_impl::{{closure}}\n             at /home/stn/tikv/compo
nents/tikv_util/src/worker/pool.rs:430\n  11: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /home/stn/.rustup/toolchains/nightly-2021-04-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:8
0\n  12: <yatp::task::future::Runner as yatp::pool::runner::Runner>::handle\n             at /home/stn/.cargo/git/checkouts/yatp-e704b73c3ee279b6/6bbea16/src/task/future.rs:261\n  13: <tikv_util::yatp_pool::YatpPoolRunner<T> as yatp::pool::runner::Runner>::handle\n
      at /home/stn/tikv/components/tikv_util/src/yatp_pool/mod.rs:93\n  14: yatp::pool::worker::WorkerThread<T,R>::run\n             at /home/stn/.cargo/git/checkouts/yatp-e704b73c3ee279b6/6bbea16/src/pool/worker.rs:48\n  15: yatp::pool::builder::LazyBuilder<T>::build::{{
closure}}\n             at /home/stn/.cargo/git/checkouts/yatp-e704b73c3ee279b6/6bbea16/src/pool/builder.rs:91\n  16: std::sys_common::backtrace::__rust_begin_short_backtrace\n             at /home/stn/.rustup/toolchains/nightly-2021-04-15-x86_64-unknown-linux-gnu/lib/rus
tlib/src/rust/library/std/src/sys_common/backtrace.rs:125\n  17: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}\n             at /home/stn/.rustup/toolchains/nightly-2021-04-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:47
4\n  18: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n             at /home/stn/.rustup/toolchains/nightly-2021-04-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:344\n  19: std::panicking::try::do_call\n
       at /home/stn/.rustup/toolchains/nightly-2021-04-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:379\n  20: __rust_try.llvm.18008344519231701248\n  21: std::panicking::try\n             at /home/stn/.rustup/toolchains/nightly-2021-04-15
-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:343\n  22: std::panic::catch_unwind\n             at /home/stn/.rustup/toolchains/nightly-2021-04-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:431\n  23: std::thread::
Builder::spawn_unchecked::{{closure}}\n             at /home/stn/.rustup/toolchains/nightly-2021-04-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:473\n  24: core::ops::function::FnOnce::call_once{{vtable.shim}}\n             at /home/stn/.
rustup/toolchains/nightly-2021-04-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227\n  25: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library
/alloc/src/boxed.rs:1546\n      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/alloc/src/boxed.rs:1546\n      std::sys::unix::thread::Thread::new::thread_start\n
 at library/std/src/sys/unix/thread.rs:71\n  26: start_thread\n  27: __clone\n"] [location=/home/stn/tikv/components/cdc/src/endpoint.rs:378] [thread_name=cdc-0]
```

</details>

### Steps to reproduce
<!-- If possible, provide a recipe for reproducing the error. A complete runnable program is good. -->

Just run CDC.

### What did you expect?

No panic.

--- a/components/cdc/Cargo.toml
+++ b/components/cdc/Cargo.toml
@@ -71,6 +71,7 @@ tempfile = "3.0"
 test_raftstore = { path = "../test_raftstore" }
 test_util = { path = "../test_util" }
 panic_hook = { path = "../panic_hook" }
+time = "0.1"
 
 [[test]]
 name = "integrations"

--- a/components/cdc/src/endpoint.rs
+++ b/components/cdc/src/endpoint.rs
@@ -342,15 +342,7 @@ impl<T: 'static + RaftStoreRouter> Endpoint<T> {
                 if is_last {
                     let delegate = self.capture_regions.remove(&region_id).unwrap();
                     if let Some(reader) = self.store_meta.lock().unwrap().readers.get(&region_id) {
-                        if let Err(e) = reader
-                            .txn_extra_op
-                            .compare_exchange(TxnExtraOp::ReadOldValue, TxnExtraOp::Noop)
-                        {
-                            panic!(
-                                "unexpect txn extra op {:?}, region_id: {:?}, downstream_id: {:?}, conn_id: {:?}",
-                                e, region_id, downstream_id, conn_id
-                            );
-                        }
+                        reader.txn_extra_op.store(TxnExtraOp::Noop);
                     }
                     // Do not continue to observe the events of the region.
                     let oid = self.observer.unsubscribe_region(region_id, delegate.id);
@@ -1148,8 +1140,10 @@ mod tests {
     use kvproto::kvrpcpb::Context;
     use raftstore::errors::Error as RaftStoreError;
     use raftstore::store::msg::CasualMessage;
+    use raftstore::store::ReadDelegate;
     use std::collections::BTreeMap;
     use std::fmt::Display;
+    use std::sync::atomic::AtomicBool;
     use std::sync::mpsc::{channel, Receiver, RecvTimeoutError, Sender};
     use tempfile::TempDir;
     use test_raftstore::MockRaftStoreRouter;
@@ -1160,6 +1154,7 @@ mod tests {
     use tikv_util::collections::HashSet;
     use tikv_util::config::ReadableDuration;
     use tikv_util::worker::{dummy_scheduler, Builder as WorkerBuilder, Worker};
+    use time::Timespec;
 
     use super::*;
     use crate::channel;
@@ -1671,6 +1666,21 @@ mod tests {
         let (task_sched, _task_rx) = dummy_scheduler();
         let raft_router = MockRaftStoreRouter::new();
         let _raft_rx = raft_router.add_region(1 /* region id */, 100 /* cap */);
+        let mut region = Region::default();
+        region.set_id(1);
+        let store_meta = Arc::new(Mutex::new(StoreMeta::new(0)));
+        let read_delegate = ReadDelegate {
+            tag: String::new(),
+            region: region.clone(),
+            peer_id: 2,
+            term: 1,
+            applied_index_term: 1,
+            leader_lease: None,
+            last_valid_ts: RefCell::new(Timespec::new(0, 0)),
+            invalid: Arc::new(AtomicBool::new(false)),
+            txn_extra_op: Arc::new(AtomicCell::new(TxnExtraOp::default())),
+        };
+        store_meta.lock().unwrap().readers.insert(1, read_delegate);
         let observer = CdcObserver::new(task_sched.clone());
         let pd_client = Arc::new(TestPdClient::new(0, true));
         let mut ep = Endpoint::new(
@@ -1679,7 +1689,7 @@ mod tests {
             task_sched,
             raft_router,
             observer,
-            Arc::new(Mutex::new(StoreMeta::new(0))),
+            store_meta,
         );
         let (tx, rx) = channel::canal(1);
         let mut rx = rx.drain();

--- a/components/raftstore/src/store/worker/read.rs
+++ b/components/raftstore/src/store/worker/read.rs
@@ -32,15 +32,15 @@ use crate::store::fsm::store::StoreMeta;
 /// A read only delegate of `Peer`.
 #[derive(Clone, Debug)]
 pub struct ReadDelegate {
-    region: metapb::Region,
-    peer_id: u64,
-    term: u64,
-    applied_index_term: u64,
-    leader_lease: Option<RemoteLease>,
-    last_valid_ts: RefCell<Timespec>,
+    pub region: metapb::Region,
+    pub peer_id: u64,
+    pub term: u64,
+    pub applied_index_term: u64,
+    pub leader_lease: Option<RemoteLease>,
+    pub last_valid_ts: RefCell<Timespec>,
 
-    tag: String,
-    invalid: Arc<AtomicBool>,
+    pub tag: String,
+    pub invalid: Arc<AtomicBool>,
     pub txn_extra_op: Arc<AtomicCell<TxnExtraOp>>,
 }
 

